src/core/exception.d

```diff
@@ -855,8 +878,51 @@ extern (C)
     }
 }
 
+private size_t alignedStoreSize()
+{
+    import core.internal.traits : maxAlignment;
+
+    // max alignment of all fields plus the hidden pointer
+    enum alignment = maxAlignment!(
+        void*,
+        typeof(ArrayIndexError.tupleof),
+        typeof(ArraySliceError.tupleof),
+        typeof(AssertError.tupleof),
+        typeof(FinalizeError.tupleof),
+        typeof(ForkError.tupleof),
+        typeof(InvalidMemoryOperationError.tupleof),
+        typeof(OutOfMemoryError.tupleof),
+        typeof(RangeError.tupleof),
+        typeof(SwitchError.tupleof),
+    );
+
+    size_t size;
+    foreach (s; [
+        __traits(classInstanceSize, ArrayIndexError),
+        __traits(classInstanceSize, ArraySliceError),
+        __traits(classInstanceSize, AssertError),
+        __traits(classInstanceSize, FinalizeError),
+        __traits(classInstanceSize, ForkError),
+        __traits(classInstanceSize, InvalidMemoryOperationError),
+        __traits(classInstanceSize, OutOfMemoryError),
+        __traits(classInstanceSize, RangeError),
+        __traits(classInstanceSize, SwitchError),
+    ])
+        if (s > size) size = s;
+
+    enum alignmentMask = alignment - 1;
+
+    // sanity check for the alignment value
+    static assert (
+            alignment > 0 && !(alignmentMask & alignment),
+            "alignment must be non-zero and power of 2");
+
+    // aligned(size) + extra alignment padding
+    return ((size + alignmentMask) & ~alignmentMask) + alignment;
+}
+
 // TLS storage shared for all errors, chaining might create circular reference
-private align(2 * size_t.sizeof) void[256] _store;
+private void[alignedStoreSize()] _store;
```

Comment:
Replacing proper alignment by increasing the allocation size? This would be a regression for LDC at least.

The problem surely isn't the approximation - for x86_64, it's 16, what the GC guarantees. So if you hit a unittest regression somewhere else after optimizing the TLS size here with DMD, that's probably a sign that another allocation hasn't been properly aligned and just happened to work because of the old round 256 size (or a DMD backend bug).

Reply:
[And AFAICT, you're still using `_store.ptr` directly, no manual alignment.]

Reply:
> Replacing proper alignment by increasing the allocation size? This would be a regression for LDC at least.
> 
> The problem surely isn't the approximation - for x86_64, it's 16, what the GC guarantees. So if you hit a unittest regression somewhere else after optimizing the TLS size here with DMD, that's probably a sign that another allocation hasn't been properly aligned and just happened to work because of the old round 256 size (or a DMD backend bug).

This is pretty much the same as this: https://github.com/ldc-developers/phobos/blob/ldc/std/typecons.d#L8165 . This is doing the alignment by hand.

I don't see the problem with this, as is now. And this is trusted to do, unless a custom alignment is done on one of those classes. And to solve it, we need the trait you mentioned above.

Reply:
Oh wow, that line is 10 years old. Back then, `align(N)` might have been buggy. There shouldn't be any reason for manual alignments anymore if the alignment is known at compile-time, as in these cases.

Reply:
I thought `void[]` had some special case. I'll change it then.

Reply:
1. AFAICT, you previously only padded the allocation, but still emplaced at `_store.ptr` (https://github.com/dlang/druntime/pull/3802#discussion_r861393064). So the actual guaranteed alignment for `_store` and its emplaced Errors was 1, success probably just a lucky punch.
2. You now don't have a nice round size anymore without the manual 'alignment' padding, which probably makes the side effect for *other* TLS allocations with insufficient alignment surface again (or a DMD backend bug). https://github.com/dlang/druntime/pull/3802#discussion_r861391531. The callstack should reveal which unittest failed, to track down the wrongly aligned allocation.

Reply:
The file/line infos for the failing macOS 11 job are clearly wrong, but the problem probably comes from https://github.com/dlang/druntime/blob/0c2400e3f4dc1baa214d22e90fc4b886d2a47cef/src/core/lifetime.d#L1201. It uses the class *ref* alignment, but 8 should still be fine for that little class; so it might be a DMD backend bug after all.

Reply:
Ah, right, I forgot to change on the emplace side :facepalm: . I think I found the side effect. The weirdest part is that it was only failing on macOS, so we probably got lucky all this time with some other introduced allocation on TLS in the other platforms.

I think now things _align_ in my brain too.

Reply:
> . It uses the class _ref_ alignment, but 8 should still be fine for that little class; so it might be a DMD backend bug after all.

Yup, probably, given that `A` only have 8-byte aligned memory. I need to dig a bit more to spot the issue, but probably someone who has macOS would be more helpful to reproduce this in a sandbox with a minimal example.

Reply:
> I need to dig a bit more to spot the issue, but probably someone who has macOS would be more helpful to reproduce this in a sandbox with a minimal example.

I've just tried to cross-compile with DMD, and then using `llvm-objdump` to inspect the Mach-O object file, for a trivial:
```D
module mod;

int a;
align(64) int b;
```

```
$ dmd -betterC -c testalign.d -target=x86_64-apple-darwin
$ llvm-objdump --section-headers testalign.o

testalign.o:	file format Mach-O 64-bit x86-64

Sections:
Idx Name          Size     VMA              Type
  0 __text        00000000 0000000000000000 TEXT
  1 __data        00000000 0000000000000000 DATA
  2 __const       00000000 0000000000000000 DATA
  3 __bss         00000000 0000000000000030 BSS
  4 __const       00000000 0000000000000000 DATA
  5 __thread_bss  00000008 0000000000000030 DATA
  6 __thread_vars 00000030 0000000000000000 DATA
```

With LDC however:
```
  1 __thread_bss  00000044 0000000000000040 DATA
```

So with DMD, the `__thread_bss` section is only 8 bytes large, and the VMA only 16-bytes aligned. With LDC, the size is 68 bytes, and the VMA is 64-bytes aligned.

With DMD for Linux x64, same 68 bytes size:
```
 14 .tbss           00000044 0000000000000000 BSS
```

Reply:
So aligned TLS on macOS is clearly not working :upside_down_face: .

Reply:
=> https://issues.dlang.org/show_bug.cgi?id=23080